dot/parachain/candidate-validation/candidate_validation_test.go

```diff
@@ -677,3 +669,190 @@ func Test_performBasicChecks(t *testing.T) {
 		})
 	}
 }
+
+func TestCandidateValidation_validateFromChainState(t *testing.T) {
+	t.Parallel()
+	candidateReceipt, validationCode := createTestCandidateReceiptAndValidationCode(t)
+	candidateReceipt2 := candidateReceipt
+	candidateReceipt2.Descriptor.PovHash = common.MustHexToHash(
+		"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
+	candidateReceipt2.Descriptor.ParaID = 2
+
+	candidateReceipt3 := candidateReceipt
+	candidateReceipt3.Descriptor.ParaID = 3
+
+	candidateReceipt4 := candidateReceipt
+	candidateReceipt4.Descriptor.ParaID = 4
+	candidateReceipt4.Descriptor.ValidationCodeHash = parachaintypes.ValidationCodeHash(common.MustHexToHash(
+		"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"))
+
+	candidateReceipt5 := candidateReceipt
+	candidateReceipt5.Descriptor.ParaID = 5
+
+	povHashMismatch := PoVHashMismatch
+	paramsTooLarge := ParamsTooLarge
+	codeHashMismatch := CodeHashMismatch
+	badSignature := BadSignature
+
+	ctrl := gomock.NewController(t)
+	t.Cleanup(ctrl.Finish)
```

Comment:
We have to clean up at the end, Right?
```suggestion
	defer t.Cleanup(ctrl.Finish)
```

Reply:
I think I can use t.Cleanup instead of defer, if I understand correctly: https://stackoverflow.com/questions/61609085/what-is-useful-for-t-cleanup/61609670#61609670